// react-native-graph/lib/commonjs/AnimatedLineGraph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnimatedLineGraph = AnimatedLineGraph;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeSkia = require("@shopify/react-native-skia");

var _SelectionDot = require("./SelectionDot");

var _CreateGraphPath = require("./CreateGraphPath");

var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));

var _getSixDigitHex = require("./utils/getSixDigitHex");

var _reactNativeGestureHandler = require("react-native-gesture-handler");

var _usePanGesture = require("./hooks/usePanGesture");

var _GetYForX = require("./GetYForX");

var _hexToRgba = require("./utils/hexToRgba");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const INDICATOR_RADIUS = 7;
const INDICATOR_BORDER_MULTIPLIER = 1.3;
const INDICATOR_PULSE_BLUR_RADIUS_SMALL = INDICATOR_RADIUS * INDICATOR_BORDER_MULTIPLIER;
const INDICATOR_PULSE_BLUR_RADIUS_BIG = INDICATOR_RADIUS * INDICATOR_BORDER_MULTIPLIER + 20; // weird rea type bug

const ReanimatedView = _reactNativeReanimated.default.View;

function AnimatedLineGraph(_ref) {
  let {
    points,
    color,
    gradientFillColors,
    lineThickness = 3,
    range,
    enableFadeInMask,
    enablePanGesture,
    onPointSelected,
    onGestureStart,
    onGestureEnd,
    SelectionDot = _SelectionDot.SelectionDot,
    enableIndicator = false,
    indicatorPulsating = false,
    horizontalPadding = enableIndicator ? INDICATOR_RADIUS * INDICATOR_BORDER_MULTIPLIER : 0,
    verticalPadding = lineThickness,
    TopAxisLabel,
    BottomAxisLabel,
    ...props
  } = _ref;
  const [width, setWidth] = (0, _react.useState)(0);
  const [height, setHeight] = (0, _react.useState)(0);
  const interpolateProgress = (0, _reactNativeSkia.useValue)(0);
  const {
    gesture,
    isActive,
    x
  } = (0, _usePanGesture.usePanGesture)({
    holdDuration: 300
  });
  const circleX = (0, _reactNativeSkia.useValue)(0);
  const circleY = (0, _reactNativeSkia.useValue)(0);
  const pathEnd = (0, _reactNativeSkia.useValue)(0);
  const indicatorRadius = (0, _reactNativeSkia.useValue)(enableIndicator ? INDICATOR_RADIUS : 0);
  const indicatorBorderRadius = (0, _reactNativeSkia.useComputedValue)(() => indicatorRadius.current * INDICATOR_BORDER_MULTIPLIER, [indicatorRadius]);
  const pulseTrigger = (0, _reactNativeReanimated.useDerivedValue)(() => {
    'worklet';

    return isActive.value ? 1 : 0;
  }, []);
  const indicatorPulseAnimation = (0, _reactNativeReanimated.useSharedValue)(0);
  const indicatorPulseRadius = (0, _reactNativeSkia.useValue)(INDICATOR_PULSE_BLUR_RADIUS_SMALL);
  const indicatorPulseOpacity = (0, _reactNativeSkia.useValue)(1);
  const positions = (0, _reactNativeSkia.useComputedValue)(() => [0, Math.min(0.15, pathEnd.current), pathEnd.current, pathEnd.current, 1], [pathEnd]);
  const onLayout = (0, _react.useCallback)(_ref2 => {
    let {
      nativeEvent: {
        layout
      }
    } = _ref2;
    setWidth(Math.round(layout.width));
    setHeight(Math.round(layout.height));
  }, []);
  const straightLine = (0, _react.useMemo)(() => {
    const path = _reactNativeSkia.Skia.Path.Make();

    path.moveTo(0, height / 2);

    for (let i = 0; i < width - 1; i += 2) {
      const x = i;
      const y = height / 2;
      path.cubicTo(x, y, x, y, x, y);
    }

    return path;
  }, [height, width]);
  const paths = (0, _reactNativeSkia.useValue)({});
  const gradientPaths = (0, _reactNativeSkia.useValue)({});
  const commands = (0, _react.useRef)([]);
  const [commandsChanged, setCommandsChanged] = (0, _react.useState)(0);
  const pathRange = (0, _react.useMemo)(() => (0, _CreateGraphPath.getGraphPathRange)(points, range), [points, range]);
  const drawingWidth = (0, _react.useMemo)(() => {
    const lastPoint = points[points.length - 1];
    return Math.max(Math.floor((width - 2 * horizontalPadding) * (0, _CreateGraphPath.pixelFactorX)(lastPoint.date, pathRange.x.min, pathRange.x.max)), 0);
  }, [horizontalPadding, pathRange.x.max, pathRange.x.min, points, width]);
  const indicatorX = (0, _react.useMemo)(() => commandsChanged >= 0 ? Math.floor(drawingWidth) + horizontalPadding : undefined, [commandsChanged, drawingWidth, horizontalPadding]);
  const indicatorY = (0, _react.useMemo)(() => commandsChanged >= 0 && indicatorX != null ? (0, _GetYForX.getYForX)(commands.current, indicatorX) : undefined, [commandsChanged, indicatorX]);
  const indicatorPulseColor = (0, _react.useMemo)(() => (0, _hexToRgba.hexToRgba)(color, 0.4), [color]);
  const shouldFillGradient = gradientFillColors != null;
  (0, _react.useEffect)(() => {
    var _previous$to2, _from$interpolate2;

    if (height < 1 || width < 1) {
      // view is not yet measured!
      return;
    }

    if (points.length < 1) {
      // points are still empty!
      return;
    }

    let path;
    let gradientPath;
    const createGraphPathProps = {
      points: points,
      range: pathRange,
      horizontalPadding: horizontalPadding,
      verticalPadding: verticalPadding,
      canvasHeight: height,
      canvasWidth: width
    };

    if (shouldFillGradient) {
      const {
        path: pathNew,
        gradientPath: gradientPathNew
      } = (0, _CreateGraphPath.createGraphPathWithGradient)(createGraphPathProps);
      path = pathNew;
      gradientPath = gradientPathNew;
    } else {
      path = (0, _CreateGraphPath.createGraphPath)(createGraphPathProps);
    }

    commands.current = path.toCmds();

    if (gradientPath != null) {
      var _previous$to, _from$interpolate;

      const previous = gradientPaths.current;
      let from = (_previous$to = previous.to) !== null && _previous$to !== void 0 ? _previous$to : straightLine;
      if (previous.from != null && interpolateProgress.current < 1) from = (_from$interpolate = from.interpolate(previous.from, interpolateProgress.current)) !== null && _from$interpolate !== void 0 ? _from$interpolate : from;

      if (gradientPath.isInterpolatable(from)) {
        gradientPaths.current = {
          from: from,
          to: gradientPath
        };
      } else {
        gradientPaths.current = {
          from: gradientPath,
          to: gradientPath
        };
      }
    }

    const previous = paths.current;
    let from = (_previous$to2 = previous.to) !== null && _previous$to2 !== void 0 ? _previous$to2 : straightLine;
    if (previous.from != null && interpolateProgress.current < 1) from = (_from$interpolate2 = from.interpolate(previous.from, interpolateProgress.current)) !== null && _from$interpolate2 !== void 0 ? _from$interpolate2 : from;

    if (path.isInterpolatable(from)) {
      paths.current = {
        from: from,
        to: path
      };
    } else {
      paths.current = {
        from: path,
        to: path
      };
    }

    setCommandsChanged(commandsChanged + 1);
    (0, _reactNativeSkia.runSpring)(interpolateProgress, {
      from: 0,
      to: 1
    }, {
      mass: 1,
      stiffness: 500,
      damping: 400,
      velocity: 0
    }); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [height, horizontalPadding, interpolateProgress, pathRange, paths, shouldFillGradient, gradientPaths, points, range, straightLine, verticalPadding, width]);
  const gradientColors = (0, _react.useMemo)(() => {
    if (enableFadeInMask) {
      return [`${(0, _getSixDigitHex.getSixDigitHex)(color)}00`, `${(0, _getSixDigitHex.getSixDigitHex)(color)}ff`, `${(0, _getSixDigitHex.getSixDigitHex)(color)}ff`, `${(0, _getSixDigitHex.getSixDigitHex)(color)}33`, `${(0, _getSixDigitHex.getSixDigitHex)(color)}33`];
    } else {
      return [color, color, color, `${(0, _getSixDigitHex.getSixDigitHex)(color)}33`, `${(0, _getSixDigitHex.getSixDigitHex)(color)}33`];
    }
  }, [color, enableFadeInMask]);
  const path = (0, _reactNativeSkia.useComputedValue)(() => {
    var _paths$current$from, _paths$current$to;

    const from = (_paths$current$from = paths.current.from) !== null && _paths$current$from !== void 0 ? _paths$current$from : straightLine;
    const to = (_paths$current$to = paths.current.to) !== null && _paths$current$to !== void 0 ? _paths$current$to : straightLine;
    return to.interpolate(from, interpolateProgress.current);
  }, // RN Skia deals with deps differently. They are actually the required SkiaValues that the derived value listens to, not react values.
  [interpolateProgress]);
  const gradientPath = (0, _reactNativeSkia.useComputedValue)(() => {
    var _gradientPaths$curren, _gradientPaths$curren2;

    const from = (_gradientPaths$curren = gradientPaths.current.from) !== null && _gradientPaths$curren !== void 0 ? _gradientPaths$curren : straightLine;
    const to = (_gradientPaths$curren2 = gradientPaths.current.to) !== null && _gradientPaths$curren2 !== void 0 ? _gradientPaths$curren2 : straightLine;
    return to.interpolate(from, interpolateProgress.current);
  }, // RN Skia deals with deps differently. They are actually the required SkiaValues that the derived value listens to, not react values.
  [interpolateProgress]);
  const stopPulsating = (0, _react.useCallback)(() => {
    (0, _reactNativeReanimated.cancelAnimation)(indicatorPulseAnimation);
    indicatorPulseAnimation.value = 0;
  }, [indicatorPulseAnimation]);
  const startPulsating = (0, _react.useCallback)(() => {
    indicatorPulseAnimation.value = (0, _reactNativeReanimated.withRepeat)((0, _reactNativeReanimated.withDelay)(1000, (0, _reactNativeReanimated.withSequence)((0, _reactNativeReanimated.withTiming)(1, {
      duration: 1100
    }), (0, _reactNativeReanimated.withTiming)(0, {
      duration: 0
    }), // revert to 0
    (0, _reactNativeReanimated.withTiming)(0, {
      duration: 1200
    }), // delay between pulses
    (0, _reactNativeReanimated.withTiming)(1, {
      duration: 1100
    }), (0, _reactNativeReanimated.withTiming)(1, {
      duration: 2000
    }) // delay after both pulses
    )), -1);
  }, [indicatorPulseAnimation]);
  const setFingerX = (0, _react.useCallback)(fingerX => {
    const lowerBound = horizontalPadding;
    const upperBound = drawingWidth + horizontalPadding;
    const fingerXInRange = Math.min(Math.max(fingerX, lowerBound), upperBound);
    const y = (0, _GetYForX.getYForX)(commands.current, fingerXInRange);

    if (y != null) {
      circleY.current = y;
      circleX.current = fingerXInRange;
    }

    if (fingerX > lowerBound && fingerX < upperBound && isActive.value) pathEnd.current = fingerX / width;
    const actualFingerX = fingerX - horizontalPadding;
    const index = Math.round(actualFingerX / upperBound * points.length);
    const pointIndex = Math.min(Math.max(index, 0), points.length - 1);
    const dataPoint = points[pointIndex];
    if (dataPoint != null) onPointSelected === null || onPointSelected === void 0 ? void 0 : onPointSelected(dataPoint);
  }, [circleX, circleY, drawingWidth, horizontalPadding, isActive.value, onPointSelected, pathEnd, points, width]);
  const setIsActive = (0, _react.useCallback)(active => {
    (0, _reactNativeSkia.runSpring)(indicatorRadius, !active ? INDICATOR_RADIUS : 0, {
      mass: 1,
      stiffness: 1000,
      damping: 50,
      velocity: 0
    });

    if (active) {
      onGestureStart === null || onGestureStart === void 0 ? void 0 : onGestureStart();
      stopPulsating();
    } else {
      onGestureEnd === null || onGestureEnd === void 0 ? void 0 : onGestureEnd();
      pathEnd.current = 1;
      startPulsating();
    }
  }, [indicatorRadius, onGestureEnd, onGestureStart, pathEnd, startPulsating, stopPulsating]);
  (0, _reactNativeReanimated.useAnimatedReaction)(() => x.value, fingerX => {
    if (isActive.value || fingerX) {
      (0, _reactNativeReanimated.runOnJS)(setFingerX)(fingerX);
    }
  }, [isActive, setFingerX, width, x]);
  (0, _reactNativeReanimated.useAnimatedReaction)(() => isActive.value, active => {
    (0, _reactNativeReanimated.runOnJS)(setIsActive)(active);
  }, [isActive, setIsActive]);
  (0, _react.useEffect)(() => {
    if (points.length !== 0 && commands.current.length !== 0) pathEnd.current = 1;
  }, [commands, pathEnd, points.length]);
  (0, _react.useEffect)(() => {
    if (indicatorPulsating) {
      startPulsating();
    } // eslint-disable-next-line react-hooks/exhaustive-deps

  }, [indicatorPulsating]);
  (0, _reactNativeSkia.useSharedValueEffect)(() => {
    if (pulseTrigger.value === 0) {
      indicatorPulseRadius.current = (0, _reactNativeSkia.mix)(indicatorPulseAnimation.value, INDICATOR_PULSE_BLUR_RADIUS_SMALL, INDICATOR_PULSE_BLUR_RADIUS_BIG);
      indicatorPulseOpacity.current = (0, _reactNativeSkia.mix)(indicatorPulseAnimation.value, 1, 0);
    } else {
      indicatorPulseRadius.current = 0;
    }
  }, indicatorPulseAnimation, pulseTrigger);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, props, /*#__PURE__*/_react.default.createElement(_reactNativeGestureHandler.GestureDetector, {
    gesture: enablePanGesture ? gesture : undefined
  }, /*#__PURE__*/_react.default.createElement(ReanimatedView, {
    style: [styles.container, styles.axisLabelContainer]
  }, TopAxisLabel != null && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.axisRow
  }, /*#__PURE__*/_react.default.createElement(TopAxisLabel, null)), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container,
    onLayout: onLayout
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Canvas, {
    style: styles.svg
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Group, null, /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Path // @ts-ignore
  , {
    path: path,
    strokeWidth: lineThickness,
    style: "stroke",
    strokeJoin: "round",
    strokeCap: "round"
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSkia.LinearGradient, {
    start: (0, _reactNativeSkia.vec)(0, 0),
    end: (0, _reactNativeSkia.vec)(width, 0),
    colors: gradientColors,
    positions: positions
  })), shouldFillGradient && /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Path // @ts-ignore
  , {
    path: gradientPath
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSkia.LinearGradient, {
    start: (0, _reactNativeSkia.vec)(0, 0),
    end: (0, _reactNativeSkia.vec)(0, height),
    colors: gradientFillColors
  }))), SelectionDot != null && /*#__PURE__*/_react.default.createElement(SelectionDot, {
    isActive: isActive,
    color: color,
    lineThickness: lineThickness,
    circleX: circleX,
    circleY: circleY
  }), enableIndicator && /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Group, null, indicatorPulsating && /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Circle, {
    cx: indicatorX,
    cy: indicatorY,
    r: indicatorPulseRadius,
    opacity: indicatorPulseOpacity,
    color: indicatorPulseColor,
    style: "fill"
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Circle, {
    cx: indicatorX,
    cy: indicatorY,
    r: indicatorBorderRadius,
    color: '#ffffff'
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Shadow, {
    dx: 2,
    dy: 2,
    color: "rgba(0,0,0,0.2)",
    blur: 4
  })), /*#__PURE__*/_react.default.createElement(_reactNativeSkia.Circle, {
    cx: indicatorX,
    cy: indicatorY,
    r: indicatorRadius,
    color: color
  })))), BottomAxisLabel != null && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.axisRow
  }, /*#__PURE__*/_react.default.createElement(BottomAxisLabel, null)))));
}

const styles = _reactNative.StyleSheet.create({
  svg: {
    flex: 1
  },
  container: {
    flex: 1
  },
  axisLabelContainer: {
    paddingVertical: 20
  },
  axisRow: {
    height: 17
  }
});
